client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,329 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Estimate of the max base RPC payload size when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload
+// does not include the RPC result.
+//
+// The estimate is based on the JSONRPC response message which has the following format:
+// `{"jsonrpc":"2.0","result":[],"id":18446744073709551615}`.
+//
+// We care about the total size of the payload because jsonrpc-server will simply ignore
+// messages larger than `sc_rpc_server::MAX_PAYLOAD` and the caller will not get any
+// response.
+const BASE_RPC_PAYLOAD: usize = 100;
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &str = "pallet,frame,state";
+const TRACE_TARGET: &str = "block_trace";
+// Default to not filtering based on storage keys as storage keys vary per chain.
+const DEFAULT_STORAGE_KEYS: &str = "";
+// The name of a field required for all events.
+const REQUIRED_EVENT_FIELD: &str  = "method";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
```

Comment:
I believe this is called for every span single span and event that the subscriber encounters (so a lot).

Yeah we need to match on partial prefix. I honestly don't have any great ideas unless we wanted to use some sort of trie that accommodates prefix matching. 

In the default scenario, targets is a constant of 3; and in other expected use cases I Imagine targets will not get too long (e.g. how often are people specifying over 10). 

Do you have ideas for how to go about benchmarking this? And considering we are still trying to ship an MVP do you think its worth doing?

Reply:
My idea wouldn't work anyway. It's not possible to use `(String, Level)` as a hash key because of [this](https://github.com/tokio-rs/tracing/issues/1376). Linear search on such a small will have to due for now.

Reply:
I don't think we should do it but we could use 2 data structures:

1) HashMap `target => level`
2) Prefix trie with targets

So first see if the event/span target is in the prefix trie. If it is a match, take the positive target from prefix trie and then key into the HasMap to get the level.

Reply:
Worth pondering but I agree we leave be for now.